### PR TITLE
Fixed experiments plugin. (for now)

### PR DESCRIPTION
**Pull request by ImperiusFate750**
Works out perfectly (for now). Fixed the weird bug where Discord would fail to upload images with this plugin enabled or it would double send an image.

Also fixed the abnormally high CPU usage that users had with this plugin enabled while looking at someone's else share screen.

### CODE REVIEW

**Comment by Tharki-God** [comment]
@ImperiusFate750  the plugin in giving the following error on disabling the plugin
image

**Comment by ImperiusFate750** [comment]
> @ImperiusFate750 the plugin in giving the following error on disabling the plugin image

That's weird. Disabling the plugin should work. You may even see the experiments tab still in there, but the experiments themselves are fully disabled as this is only a visual bug. There may be something wrong with your client, I can't reproduce this error.

Are you sure you're using the latest code? 
https://raw.githubusercontent.com/Inve1951/BetterDiscordStuff/606f8bfbdf9ba8df53b32773ef208088baf89877/plugins/discordexperiments.plugin.js

**Comment by Tharki-God** [comment]
yes i am on the latest code and i have tried with two clients and same error

**Comment by rollingmoai** [comment]
This fix works and I am not getting the error that @Tharki-God is getting.

**Comment by ImperiusFate750** [comment]
> This fix works and I am not getting the error that @Tharki-God is getting.

I'm happy to hear it.

**Comment by CJStyles123** [comment]
Seems like plugin broke again.
image
image

**Comment by ImperiusFate750** [comment]
> Seems like plugin broke again. image image

Unfortunately yes, mine broke too. We'll have to wait a bit for a fix

File: plugins/discordexperiments.plugin.js [hunk 1]
```diff
@@ -2,27 +2,30 @@
...
 module.exports = class {
 	getName(){ return "Discord Experiments"; }
 
 	start() {
-    storeExports.default = new Proxy(original, {
-			get(_, key) {
-				return key === "isDeveloper" ? true : original[key];
-			}
-		});
+		const nodes = Object.values(settingsStore.default._dispatcher._dependencyGraph.nodes);
```
Comment (dav1312) [inline plugins/discordexperiments.plugin.js:18]:
```js
const nodes = Object.values(settingsStore.default._dispatcher._actionHandlers._dependencyGraph.nodes);
```

**Review by dav1312 — CHANGES_REQUESTED** [review]

File: plugins/discordexperiments.plugin.js (hunk 1 above)
Comment (Aholicknight) [reply plugins/discordexperiments.plugin.js:18]: DiscordPTB_0075086092

Still does not work

**Comment by Aholicknight** [comment]
DiscordPTB_0075086092

@dav1312 still does not work with your changes

File: plugins/discordexperiments.plugin.js (hunk 1 above)
Comment (Aholicknight) [reply plugins/discordexperiments.plugin.js:18]: @dav1312 Pull request #189 combines your change.

**Comment by Tharki-God** [comment]
i tried on virtual machine and i still get the error, mentioned above by me, after even after applying dav1312's fix

**Comment by Aholicknight** [comment]
> i tried on virtual machine and i still get the error, mentioned above by me, after even after applying dav1312's fix

@Tharki-God It works for me fine, did you use his requested change?

**Comment by Tharki-God** [comment]
> > i tried on virtual machine and i still get the error, mentioned above by me, after even after applying dav1312's fix
> 
> @Tharki-God It works for me fine, did you use his requested change?

yes, enabled experiments are , "april fool" and "powermode"

**Comment by Aholicknight** [comment]
> yes, enabled experiments are , "april fool" and "powermode"

@Tharki-God Take a screenshot of this because I don't see that lol

**Comment by Mysteriouslymythical** [comment]
Works on normal discord.  Anyone have a new betterdiscord plugin for this?

**Comment by Soulivan71** [comment]
There is a fix ?

**Comment by Aholicknight** [comment]
> There is a fix ? For the new update discord

image

**Comment by CJStyles123** [comment]
> There is a fix ? For the new update discord

Discord switched to a new compiler which was meant to improve performance and speed up compiling, all of the internal code has been minified so all Plugins and snippets are patched. it's gonna take weeks if not months to fix.

**Comment by Soulivan71** [comment]
With last update when I turn off disccordexperiments all is resetting do u have a fix ?

**Comment by Soulivan71** [comment]
Discord experiments could not be started : 
> image
Can someone send me the file please

**Comment by Soulivan71** [comment]
@ImperiusFate750 Hi do you have a working file for discord experiments ?

**Comment by ImperiusFate750** [comment]
> @ImperiusFate750 Hi do you have a working file for discord experiments ?

Nope, been trying for awhile, but nothing..

**Comment by Tharki-God** [comment]
```
const {Utilities, DiscordModules: {UserStore, Dispatcher}} = ZLibrary
          const ExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "ExperimentStore" && n.actionHandler["CONNECTION_OPEN"]);
          const DeveloperExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "DeveloperExperimentStore");
          const CurrentUser = UserStore.getCurrentUser();
          ExperimentStore?.actionHandler.CONNECTION_OPEN({
            type: "CONNECTION_OPEN", user: { flags: CurrentUser.flags |= 0 }, experiments: [],
          });
          DeveloperExperimentStore.actionHandler.CONNECTION_OPEN();
          CurrentUser.flags &= ~0;
```

**Comment by Soulivan71** [comment]
> const {Utilities, DiscordModules: {UserStore, Dispatcher}} = ZLibrary
>           const ExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "ExperimentStore" && n.actionHandler["CONNECTION_OPEN"]);
>           const DeveloperExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "DeveloperExperimentStore");
>           const CurrentUser = UserStore.getCurrentUser();
>           ExperimentStore?.actionHandler.CONNECTION_OPEN({
>             type: "CONNECTION_OPEN", user: { flags: CurrentUser.flags |= 0 }, experiments: [],
>           });
>           DeveloperExperimentStore.actionHandler.CONNECTION_OPEN();
>           CurrentUser.flags &= ~0;

Only this ?

**Comment by Tharki-God** [comment]
> > const {Utilities, DiscordModules: {UserStore, Dispatcher}} = ZLibrary
> >           const ExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "ExperimentStore" && n.actionHandler["CONNECTION_OPEN"]);
> >           const DeveloperExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "DeveloperExperimentStore");
> >           const CurrentUser = UserStore.getCurrentUser();
> >           ExperimentStore?.actionHandler.CONNECTION_OPEN({
> >             type: "CONNECTION_OPEN", user: { flags: CurrentUser.flags |= 0 }, experiments: [],
> >           });
> >           DeveloperExperimentStore.actionHandler.CONNECTION_OPEN();
> >           CurrentUser.flags &= ~0;
> 
> This isn't working for me
> 

Sorry make it 1 instead of 0 at both places. I was testing a little and sent the wrong code

**Comment by Soulivan71** [comment]
Thx @Tharki-God

**Comment by Tchiemny** [comment]
> > > const {Utilities, DiscordModules: {UserStore, Dispatcher}} = ZLibrary
> > > const ExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "ExperimentStore" && n.actionHandler["CONNECTION_OPEN"]);
> > > const DeveloperExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "DeveloperExperimentStore");
> > > const CurrentUser = UserStore.getCurrentUser();
> > > ExperimentStore?.actionHandler.CONNECTION_OPEN({
> > > type: "CONNECTION_OPEN", user: { flags: CurrentUser.flags |= 0 }, experiments: [],
> > > });
> > > DeveloperExperimentStore.actionHandler.CONNECTION_OPEN();
> > > CurrentUser.flags &= ~0;
> > 
> > 
> > This isn't working for me
> 
> Sorry make it 1 instead of 0 at both places. I was testing a little and sent the wrong code

Do you have full code for this plugin?

**Comment by Tharki-God** [comment]
> > > > const {Utilities, DiscordModules: {UserStore, Dispatcher}} = ZLibrary
> > > > const ExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "ExperimentStore" && n.actionHandler["CONNECTION_OPEN"]);
> > > > const DeveloperExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "DeveloperExperimentStore");
> > > > const CurrentUser = UserStore.getCurrentUser();
> > > > ExperimentStore?.actionHandler.CONNECTION_OPEN({
> > > > type: "CONNECTION_OPEN", user: { flags: CurrentUser.flags |= 0 }, experiments: [],
> > > > });
> > > > DeveloperExperimentStore.actionHandler.CONNECTION_OPEN();
> > > > CurrentUser.flags &= ~0;
> > > 
> > > 
> > > This isn't working for me
> > 
> > Sorry make it 1 instead of 0 at both places. I was testing a little and sent the wrong code
> 
> Do you have full code for this plugin?

Make the plugin yourself

**Comment by Tchiemny** [comment]
> > > > > const {Utilities, DiscordModules: {UserStore, Dispatcher}} = ZLibrary
> > > > > const ExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "ExperimentStore" && n.actionHandler["CONNECTION_OPEN"]);
> > > > > const DeveloperExperimentStore = Utilities.findInTree(Dispatcher, n => n?.name == "DeveloperExperimentStore");
> > > > > const CurrentUser = UserStore.getCurrentUser();
> > > > > ExperimentStore?.actionHandler.CONNECTION_OPEN({
> > > > > type: "CONNECTION_OPEN", user: { flags: CurrentUser.flags |= 0 }, experiments: [],
> > > > > });
> > > > > DeveloperExperimentStore.actionHandler.CONNECTION_OPEN();
> > > > > CurrentUser.flags &= ~0;
> > > > 
> > > > 
> > > > This isn't working for me
> > > 
> > > 
> > > Sorry make it 1 instead of 0 at both places. I was testing a little and sent the wrong code
> > 
> > 
> > Do you have full code for this plugin?
> 
> Make the plugin yourself

I don't know what i do wrong

**Comment by Soulivan71** [comment]
image
Im having issue on start ^^

''Discord experiments could not be started"
Canary 219456 (8287abc) Host 1.0.73 (35715) Windows 10 64-Bit (10.0.19045)

**Comment by BossDaily** [comment]
Just use vencord, it has all the plugins you need, including a working experiments plugin

**Comment by Tharki-God** [comment]
> Just use vencord, it has all the plugins you need, including a working experiments plugin

haha replugged better

**Pull request closed without merging**